Add per-font size adjustment, don't adjust nf symbol font or emoji font; use non-Mono symbols nerd font

danielwe opened the pull request:
This adds functionality for choosing different normalization metrics for each fallback font. It's not exposed as a config option, but could be in the future, which would probably go a long way towards addressing concerns like #7929.

The currently available reference metrics are, in priority order: `ic_width, ex_height, cap_height, line_height, em_size`. The default value is `ic_width`.

By priority order, I mean that if the chosen metric is not defined in the fallback font, we move to the next metric in the list---we don't normalize by an estimated metric from the fallback font (however, we're happy to use an estimated metric from the primary font, that's how `ic_width` normalization between CJK and Latin fonts work). This extends the pattern that was used between `ic_width` and `ex_height` in the existing hardcoded rule. `line_height` is always defined, so the buck stops there.

What motivated me to implement this was the fact that, with the existing hardcoded rule, the embedded symbols-only Nerd Font was always scaled up by a factor of 1.2, which turned out to be an important reason why it's been difficult to make icon scaling work to everyone's satisfaction. Accordingly, the symbols-only font is the first to take advantage of the new functionality. If this PR is merged, #7917 is no longer needed. (To limit the scope of this PR, it only includes the minimal changes to let icon scaling take advantage of this functionality. I may submit a follow-up PR with some further icon scaling improvement enabled by this.)

danielwe committed:
Move face metric fallback estimates to the FaceMetric struct

danielwe committed:
Use em size as nerd font reference metric

danielwe commented:
Was working on adding some tests and realized that this implementation doesn't play well with things like `Collection.setSize`. Hold on while I take care of that.

danielwe committed:
Make size normalization reference customizable per face

danielwe committed:
Add unwrapConst, avoid constCast

danielwe committed:
Calculate scaled size directly, eliminate redundant resizes

danielwe committed:
Update a straggling name and signature

danielwe commented:
Just discovered from investigating #8054 that some fonts specify huge ascent and descent, adding up to a line height of almost 3 em. The current heuristics (both before and after this PR) result in super tiny glyphs when such a font is loaded as a fallback next to a "normal" font where the line height is in the neighborhood of 1 em, due to the capping of the line height ratio at 1.2. My takeaway from this is that line height is an unreliable metric that should not be used in these calculations at all. We could cap the ratio of point sizes at 1.2 instead. How does that sound?

danielwe commented on src/font/SharedGridSet.zig:
```diff
@@ -324,7 +327,7 @@ fn collection(
                 self.alloc,
                 .regular,
-                .{ .fallback_deferred = face },
+                .init(.{ .fallback_deferred = face }),
```
We probably want to opt out of scaling for the emoji font too? (Not gonna push the change right now as I know you're in the middle of reviewing.)

```suggestion
                .initWithScaleReference(.{ .fallback_deferred = face }, .em_size),
```

qwerasd205 committed:
cleanup

A variety of naming, commenting, and formatting improvements + a few
explicit error sets. This commit has no functional changes, though it
does remove a couple functions that didn't really need to exist.

qwerasd205 committed:
font: revert switch to non-mono symbols nerd font

This change might be good, but it is incomplete and not relevant to the
PR that it's a part of. I'll explore making this change separately after
this, since it might be a good idea.

(It's incomplete since the attribute data was not re-generated based on
the non-mono file.)

qwerasd205 commented:
Did some cleanup I wanna add before merging this, but it ended up being more of a change than expected. I'll finish it tomorrow most likely.

qwerasd205 committed:
font: allow fractional pixel sizes

qwerasd205 committed:
font: clean up Collection api somewhat

Move size adjustment logic out of `Entry`, I understand the impulse to
put it there but it results in passing a lot of stuff around which isn't
great.

Rework `add(...)` in to `add(...)` and `addDeferred(...)`, faces are
passed directly now instead of passing an entry, and an options struct
is used instead of positional arguments for things like style, fallback,
and size adjustment.

Change size adjustment test back to a half pixel tolerance instead of 5%
because the previous commit (allowing fractional pixel sizes) fixed the
root cause of large differences.

qwerasd205 committed:
font: use non-mono symbols nerd font for embedded symbols

I changed my mind, this is a pretty small change and relevant to the
intent of the PR. This brings the appearance of the embedded symbols
much closer to patched fonts.

With this, the sizes of most symbols are nearly identical to a patched
font, the only big difference is positioning (and TBH I think we do a
better job positioning than the patcher does, since we have knowledge
about the cell size).

danielwe commented:
Looks great, thanks for all the improvements! I also had some concerns about consistency across multiple specified fonts and especially synthesized styles, so thanks for straightening out all that.

My only question is whether the line height should be in the fallback path at all? Might be safer to just use no scaling as the catch-all fallback.

> a big aspect of that turns out to be just using the non-Mono symbols font, rather than any of these other changes.

I agree that the non-mono symbols font is much nicer; there are a lot of compromises in relative icon sizes in the mono one. However, an important condition for the non-mono version to work well is that it's loaded at the same point size as the primary. If you load it oversized, most icons have to be scaled down to the constraints, leading to a lot of compromises, while if you load it at the nominal size, most icons pass through the constraints unchanged except wrt. alignment (unless your primary font is very narrow and tall). So I'd say the magic is in the sum of these two changes.

> happier with the nerd font sizing (though I might fiddle with it a little more)

I have a branch with nerd font scaling tweaks that's ready to go once rebased. It implements a few details from font_patcher that are not implemented here, mostly regarding single vs. double-width constraints and group vs. individual alignment. What would you like me to do with this proposal? Should I submit a PR, start a discussion, or just send you a link to the branch to look at? Or should I wait and let you fiddle until you're done, and then see if I still have changes I'd like to propose?

mitchellh reviewed (approved):
Excellent work to you both! I scanned through it looking for any obvious confusion and it all looks straightforward to me, though I'm going to trust on the basic logic of size adjustment.
